File: core/documents/document_structure_engine.py
from __future__ import annotations

from typing import Any, Dict, List

MAX_SECTIONS = 5000

HEADING_PREFIXES = (
    "#",
    "##",
    "###",
)


def build_document_structure(
    text: str,
) -> Dict[str, Any]:
    sections = []

    current = {
        "title": "root",
        "content": [],
    }

    lines = text.splitlines()

    for line in lines[:MAX_SECTIONS]:
        stripped = line.strip()

        if stripped.startswith(HEADING_PREFIXES):
            if current["content"]:
                sections.append(current)

            current = {
                "title": stripped.lstrip("#").strip(),
                "content": [],
            }
        else:
            current["content"].append(stripped)

    if current["content"]:
        sections.append(current)

    return {
        "sections": sections,
        "bounded": True,
    }

File: src/documents/documentStructureEngine.ts
/**
 * Converted from Python: core/documents/document_structure_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_SECTIONS: any = 5000;
export let HEADING_PREFIXES: any = ["#", "##", "###"];
export function buildDocumentStructure(text: any): any {
  var sections: any[] = [];
  var current: any = {"title": "root", "content": []};
  var lines: any = py.splitlines(text);
  var line: any;
  for (line of py.iter(py.slice(lines, null, MAX_SECTIONS))) {
    var stripped: any = py.strip(line);
    if (py.truthy(py.startswith(stripped, HEADING_PREFIXES))) {
      if (py.truthy(py.at(current, "content"))) {
        py.listAppend(sections, current);
      }
      current = {"title": py.strip(py.lstrip(stripped, "#")), "content": []};
    } else {
      py.listAppend(py.at(current, "content"), stripped);
    }
  }
  if (py.truthy(py.at(current, "content"))) {
    py.listAppend(sections, current);
  }
  return {"sections": sections, "bounded": true};
}
